### PR TITLE
Add save data upgrade for 1.7.0.0

### PR DESCRIPTION
**Pull request by devn1x**
Until now, I stumbled over the request for a save data upgrade to 1.7.0.0 multiple times. I looked into this, because I had this 'issue' myself and wanted to try to tackle it.

In the end, I wrote this upgrade class for a save game of mine. Maybe it is of use to this project? So, if someone with a little bit more knowledge in this project could review this or give me any kind of feedback, that would be nice. Maybe I missed something, that didn't occur in my save.

Also, I prepared this to be merged into master. Is this fitting or should I merge in some other branch?

### CODE REVIEW

**Comment by dartasen** [comment]
Thank's @devn1x !

We didn't make this upgrade possible since some base data are lost in the process. However it might be better than nothing.
I don't think it will work with all bases but it's worth the try.

> Also, I prepared this to be merged into master. Is this fitting or should I merge in some other branch?
Maybe you should target the 1.7.0.0 branch, master is currently under heavy modification to allow compatiblity with the latest subnautica version

File: NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                JToken typeToken = piece["RotationMetadata"]["value"]["$type"];
+                switch (typeToken.ToString())
+                {
+                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.AnchoredFaceRotationMetadata, NitroxModel-Subnautica":
+                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.AnchoredFaceBuilderMetadata, NitroxModel-Subnautica";
+                        break;
+                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.CorridorRotationMetadata, NitroxModel-Subnautica":
+                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.CorridorBuilderMetadata, NitroxModel-Subnautica";
+                        break;
+                }
```
Comment (DaleW42) [inline NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs:48]: Need to add:
```
                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.BaseModuleRotationMetadata, NitroxModel-Subnautica":
                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.BaseModuleBuilderMetadata, NitroxModel-Subnautica";
                        break;
                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.MapRoomRotationMetadata, NitroxModel-Subnautica":
                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.MapRoomBuilderMetadata, NitroxModel-Subnautica";
                        break;
```

File: NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+                JToken typeToken = piece["RotationMetadata"]["value"]["$type"];
+                switch (typeToken.ToString())
+                {
+                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.AnchoredFaceRotationMetadata, NitroxModel-Subnautica":
+                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.AnchoredFaceBuilderMetadata, NitroxModel-Subnautica";
+                        break;
+                    case "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.CorridorRotationMetadata, NitroxModel-Subnautica":
+                        typeToken = "NitroxModel_Subnautica.DataStructures.GameLogic.Buildings.Rotation.Metadata.CorridorBuilderMetadata, NitroxModel-Subnautica";
+                        break;
```
[the remark this answers is not in the document]
Comment (devn1x) [reply NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs:47]: Thanks for pointing that out. I will  add it later today.

**Comment by devn1x** [comment]
> Thank's @devn1x !
> 
> We didn't make this upgrade possible since some base data are lost in the process. However it might be better than nothing. I don't think it will work with all bases but it's worth the try.
> 
> > Also, I prepared this to be merged into master. Is this fitting or should I merge in some other branch?
> > Maybe you should target the 1.7.0.0 branch, master is currently under heavy modification to allow compatiblity with the latest subnautica version
Thanks, I really appreciate your feedback. Would you be able to explain to me, which edge cases could prevent a successful save upgrade for the bases?

File: NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (devn1x) [reply NitroxServer/Serialization/SaveDataUpgrades/Upgrade_V1700.cs:53]: I have added the suggested cases to my code. Thank you.

**Comment by dartasen** [comment]
Bases are not 100% persistent even in 1.7.1.0, The new format introduced some data loss making the 1:1 conversion almost impossible.
1.8 is likely to break everything once again, I'm closing it for now.

Thank's for the contribution though, we appreciate it.

**Pull request closed without merging**